### ISSUE TITLE
[css-grid-3] Make align-self/justify-self align including adjacent extra space

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#10275]
This came up earlier in discussions with @jensimmons and again when she was creating demos: spanning items create gaps above them, and it would be useful to be able to align within those gaps. In particular, applying `stretch` or `end` alignment would improve the effects in demos such as https://webkit.org/demos/grid3/photos/ (e.g. in "Every 5th item").

The proposal is to use the self-alignment properties in the masonry axis to align an item above a spanner or at the end of the masonry grid in the space allotted to it + the subsequent gap.

**Comment by tabatkins** [comment]
Oh yeah, that makes sense. You *need* it to work like this, in fact - you can't align the contents of the column as a whole, *because* spanning items cross columns. So it needs to align each contiguous run of single-column items.

Does this then imply that spanning items can *never* be aligned in that axis? We don't otherwise meaningfully distinguish between a masonry of single-column items and an identical masonry with double the half-width columns and all items being two-column spanners, but I don't think I'm *opposed* to making that distinction.

**Comment by bfgeek** [comment]
Generally supportive of the stretch capability.

Two points:
 - When defining stretch needs to be explicit on how percentages resolve. E.g. potentially want flex behaviour here.
 - Need to be careful for the other non-stretch values, e.g. it runs into the same "ordering becomes unclear" issue, but perhaps less so here.

**Comment by tabatkins** [comment]
By "flex behavior" you mean "treat %s as indefinite if they're resolving against an indefinite dimension" (and stretching doesn't make a dimension become definite).

**Comment by Alimsayed** [comment]
_**[](url)**_

**Comment by fantasai** [comment]
@tabatkins I think you're confused... this issue is about aligning only the last item adjacent to the gap above spanner, not the entire stack of items above the spanner. I agree that aligning the entire stack could be useful, too, but it runs into the problems @bfgeek outlined for https://github.com/w3c/csswg-drafts/issues/8207 which is why that feature was removed.

(I thought it was a weird request at first, but then seeing Jen explaining why she needed it in her demos, it started to make sense why it would be useful to do this.)

**Comment by jensimmons** [comment]
<img width="1475" alt="grid3-alignment-1" src="https://github.com/w3c/csswg-drafts/assets/108474/cc7cfde9-67f5-45b4-8b0f-ee4958192ecb">
<img width="1475" alt="grid3-alignment-2" src="https://github.com/w3c/csswg-drafts/assets/108474/ccc7443e-e2fc-424d-8c85-5a069be0891d">
<img width="1475" alt="grid3-alignment-3" src="https://github.com/w3c/csswg-drafts/assets/108474/4c878cfc-e8b8-4ef9-a44e-eff65a2e91b1">
<img width="1475" alt="grid3-alignment-4" src="https://github.com/w3c/csswg-drafts/assets/108474/b69a8acb-a5a6-48f8-a0c5-ecb68ac6ba1b">
<img width="1377" alt="grid3-alignment-5" src="https://github.com/w3c/csswg-drafts/assets/108474/c9be509b-7de9-44d8-a980-11bf3777cc3c">

**Comment by tabatkins** [comment]
Oh, huh. *Only* aligning the last one seems fairly odd to me.

Jen's examples look good and the use-cases are definitely reasonable, but all the images have  only *one* element being affected, so it doesn't demonstrate the difference in question. If there were multiple elements, especially if there were 3+, I think it would look odd if all it did was change it from "the entire stack of elements is start-aligned" to "the entire stack of elements is start-aligned except for the last one, which is end-aligned". This seems extra true for `center`. (`stretch` might look okay, tho.)

> I agree that aligning the entire stack could be useful, too, but it runs into the problems @bfgeek outlined for https://github.com/w3c/csswg-drafts/issues/8207 which is why that feature was removed.

No, in that issue @bfgeek was objecting to the ability to align the contents of each track *independently*  - giving one track start alignment, another end alignment, etc. He recognized the value in being able to align *all* the tracks in a given direction: "The use-cases I've seen here typically align masonry axis tracks together rather than individually." and the following paragraphs in his comment.

Non-start alignments *applied to the whole masonry* can somewhat misalign the reading order from the dom order, but generally not to an enormous degree (and generally no more than what `masonry-threshold` can do).

**Comment by tabatkins** [comment]
> They could use align-self: start on enough items to fill the first row, and use align-self: end on the rest of the items to get this result.

Ooh tho, this case does look pretty good, but isn't addressed by using `align-content` as I suggest. But it still hits the same issues I mentioned, where if the later picture (of the hallway/stairwell) wasn't alone in its stack, but rather was one of three or more images between the two spanners, having just the bottom one be end-aligned would look odd. It also means you have to guess at the number of tracks you'll get, which isn't generally safe since we think auto-fill masonrys will be common.

This might argue, then, for a masonry-specific alignment property, which can distinguish between stacks which are touching spanners on only one side, and those actually wedged between spanners on both sides.

An additional plus of that approach is we can then keep `align-content` consistent with Grid, where it aligns the masonry stack as a whole, and `align-self` consistent with all the layout modes, where it does nothing because you can't align individual elements when they're laid out in a group like that.

**Comment by fantasai** [comment]
@tabatkins I think you're right that this needs the ability to select across what's at the top vs bottom of a stack of items. However I do think that adjusting just the one item, and not the whole stack, is reasonable here. You don't want the whole layout to shift as you're adding items, for example, but you might want to have the bottom row of items stretch to create a flat line--until more content loads in after it.

I think the way to do this would be with `align-items`, and following the `text-align` longhands pattern (`text-align-all`/`text-align-last`). Like this:

- align-self aligns a masonry item within the space it occupies plus the adjacent gap (if any), as [described above](https://github.com/w3c/csswg-drafts/issues/10275#issuecomment-2100950232) by @jensimmons 
- align-items-first applies to items in the first row (of grid/masonry/flex).
- align-items-last applies to items in the last row (of grid/masonry/flex).
- align-items-all applies to all the other items.
- align-items is a shorthand for all three.
- align-self overrides align-items, as usual.

**Comment by tabatkins** [comment]
Hm, I don't think that *quite* works, but it's close. In particular, if you wanted to avoid ragged edges *or* holes, and only wanted to adjust the first/last items, then you'd want `stretch` on the first/last ones, and then `end` on the non-first items in each of the first stacks, but `start` on the non-last items in the last stacks. And I'm not sure which behavior you'd want for stacks in the middle, that aren't touching the top or bottom of a masonry track.

I think we need to play with the value space some more before we hit on the right answer.

**Comment by ethanjv** [comment]
Doing some analysis of the issue, we believe that whatever behavior is expected from the alignment properties it can be accomplished by aligning _stacks of items_ as a whole.

Just to confirm that we share the same understanding, we consider that a _stack of items_ is a group of all contiguous single-track items contained between the boundaries of either the masonry container or an item spanning multiple tracks. In the following figure, we highlight all different stacks of items in pink.

![stack-example](https://github.com/user-attachments/assets/ad6c561d-d37f-4898-afea-a3863bd5725f)

Regardless of whether we want to align the stacks as a whole or only the first/last item within them, keeping track of these stacks is required in some capacity and it shouldn't shift layout.

On the other hand, we believe that allowing self-alignment outside of these stacks will be problematic and should be avoided. Consider the following example:

![opposing-alignment](https://github.com/user-attachments/assets/691a8157-36a1-41f7-bc73-30b6312c4fd1)

The pink and blue items have opposing alignment properties which have a non-intuitive interaction with each other. It only gets worse if we add more items to the stack _pushing_ each other.

We think that aligning the stacks as a whole is a superset of aligning the first/last item in a stack and should suffice to address the most interesting use case of avoiding holes between items or ragged edges at the end of the stacking axis. 

In order to avoid ragged edges at the end, one could expect align-items-last to control the behavior of the last stack in each track. Note that Tab’s proposal of keeping everything aligned as start, while making only the last item in the stack stretch is a definable behavior for the whole last stack. However, we also picture an alternate version that stretches all items in the stack proportionally:

![alignment-proposal](https://github.com/user-attachments/assets/1c343e4b-e8f3-4e35-a0ba-eebe4d617cec)

Another example, avoiding holes using the align-items-all for stacks in the middle:

![alignment-all-proposal](https://github.com/user-attachments/assets/be4f84d6-4acc-4c6c-a181-0cc29e8c7071)

Finally, without more concrete use cases for other property values, we don’t see how center or end make sense. We believe we only need to define some stretching capability to address gaps and revisit whenever is necessary.

**Comment by alisonmaher** [comment]
Theoretically, I am inclined to believe that self-alignment here should be more similar to flexbox in that it only applies in the grid axis given that masonry does not create distinct cells like grid does.

Ignoring this, what Ethan describes seems relatively doable. However, given that we allow dense packing in masonry, this means we'd need to wait to apply self-alignment to items until all items have been placed, as opposed to as soon as a spanner is added that creates space above it.

This is possible, though, if you kept track of the item stacks with open spaces as you perform layout and then apply alignment to those at the end if applicable, but this seems somewhat unfortunate, and this could also lead to layout shifts as new items are added if both self-alignment and dense packing are applied.

I suppose one way to improve that situation is if you make self-alignment in the masonry axis not apply if dense packing is also applied. However, if an author truly didn't want any extra spaces, there are scenarios where a space that cannot be filled by alignment and would need dense packing or item reordering to fill the space:

<img width="218" height="338" alt="Image" src="https://github.com/user-attachments/assets/16493ad2-a73e-4f76-896c-8c8b6ff2c33d" />

For example, in this image, you can fill the space after the first item by stretching it (likely not with dense packing), but cannot fill the remaining open space near the bottom without another single spanning item. So, I can see use cases where an author *might* want both applied at the same time.

In either case, I agree that stretch should be sufficient to handle cases where there are open spaces to fill, and the simplest approach would be to only apply alignment to the last item in a stack, that way when you add a spanner, you just need to look at the item above it in each track it spans, as opposed to all item stacks per track it spans.

**Comment by SebastianZ** [comment]
Just a note about the empty meeting notes above, the discussion of this issue was postponed on the call. That's why it's empty.

Sebastian

**Comment by alisonmaher** [comment]
As noted in https://github.com/w3c/csswg-drafts/issues/12918, items with negative margin raises questions for this issue, as well. For example, in such cases, which item "belongs" to the empty space and has alignment applied to it? When items have weird margins, this becomes unclear.

For example:

<img width="587" height="862" alt="Image" src="https://github.com/user-attachments/assets/0cdf6cf1-0f7c-4a81-a2c4-716b7e3c8d5d" />

Or

<img width="601" height="512" alt="Image" src="https://github.com/user-attachments/assets/1f62e696-3268-4119-923d-8e52c52d5231" />

This leads me back to my original thinking: "I am inclined to believe that self-alignment here should be more similar to flexbox in that it only applies in the grid axis given that masonry does not create distinct cells like grid does."

**Comment by fantasai** [comment]
@alisonmaher If the author is using negative margins, they're asking for overlap. In that case, we should give them the overlap they asked for.

**Comment by alisonmaher** [comment]
@fantasai In other words, the last item placed in a gap should be the alignment subject even if it overlaps a potentially earlier gap due to negative margin. I suppose that may be fine, just slightly strange

**Comment by tabatkins** [comment]
Yup, that's how we always treat negative margins. Even if they induce so much overlap that the last item is *fully before* a preceding item, the last item in logical order is still the last item.

**Comment by alisonmaher** [comment]
Gotcha, that makes sense - this would require just a chunk of extra bookkeeping. Essentially, if there is any item with alignment in the masonry axis, we'd need to keep track of all gaps created by spanners (similar to dense packing), as well as the last item placed in each stack above the spanner that is associated with that gap. And then apply alignment to those potential items at the end if applicable. That should be doable, just some extra mappings.

**Comment by alisonmaher** [comment]
One minor note on potential wording if we do decide to add this functionality to the spec is that we'll need to be careful when we say "the last item in the stack" can take up extra space for alignment because there are cases where the last item in a stack can't actually align because it spans a column that is blocked:

<img width="195" height="226" alt="Image" src="https://github.com/user-attachments/assets/8bbb3cf7-58da-465d-ba0d-3dade1421a01" />

For example, if we were to remove the blue item, the spanner above it could have `align-self: stretch` applied, but with the blue item blocking the spanner in the first track, it won't be able to fill the open space in the second track, so we'll somehow need to account for that case in the way this ends up being spelled out.

**Comment by tabatkins** [comment]
Okay so per resolution:

`align-self` can be used in Masonry, aligning an item in the space between its top edge and the top edge of the next item in the same track(s). This, effectively, only has an effect on the "last" item in a given stack before a later spanner makes clear space.

`align-self:normal` resolves to start or end, depending on the fill direction of the tracks.

`align-items` works as normal.

**Comment by alisonmaher** [comment]
@tabatkins I don't think we fully resolved on all details (per "RESOLVED: Find a solution to this"), like which keywords apply (all of them, or only a subset like `start`, `stretch` and `end`). We also didn't resolve on any of the details of the item(s) it would apply to, but the spec text looks reasonable since I think we'd only want to adjust the last item in a stack.

The only caveat being that I think it would be good to add some clarification about spanning items, and how there has to be space below all tracks an item spans (if the last item in a stack happens to be a spanner) as noted in https://github.com/w3c/csswg-drafts/issues/10275#issuecomment-3453544671 (but maybe that's implied)

**Comment by alisonmaher** [comment]
Adding back to the agenda to confirm what alignment keywords would apply in this case.

**Comment by alisonmaher** [comment]
If we allow `end` and `center` alignment, for example, what would happen in the following case:

<img width="191" height="206" alt="Image" src="https://github.com/user-attachments/assets/388d3f14-b204-4fc1-835e-8ddc29a90472" />

I presume the alignment-available gap below the blue item at the top won't grow when the spanner below it is also aligned to the center or end of its own gap?
